[Agent Skills] Add agent skills and rules for ECS PR Triage (#2608)
elastic/ecs: .agents/skills/ecs-rfc-guide/SKILL.md

description: >-
  Guides contributors through the Elastic Common Schema (ECS) RFC (Proposal)
  process: template sections, target maturity (alpha/beta), rfcs/text artifacts,
  and OTel alignment. Use when a change needs an RFC, when drafting or
  reviewing RFC PRs, or when the user asks how to propose new ECS field sets or
  substantial schema changes.
---

# ECS RFC (Proposal) guide

## When this applies

Use after **ecs-pr-triage** (or equivalent judgment) says **Needs RFC**, or when the user is starting a **new field set**, **breaking** change, **novel use case**, or **ECS-wide** design.

Authoritative process: [rfcs/PROCESS.md](../../../rfcs/PROCESS.md). Template: [rfcs/0000-rfc-template.md](../../../rfcs/0000-rfc-template.md). High-level triggers: [rfcs/README.md](../../../rfcs/README.md).

## Current process (short)

1. Single **Proposal** stage — template must keep `Stage: **Proposal**`.
2. Contributor opens a **PR** that adds the RFC markdown under `rfcs/` (name like `0000-<dash-separated-name>.md` until numbered).
3. Specify **Target maturity:** `alpha`, `beta`, or `mixture` (see [Field stability](../../../docs/reference/ecs-principles-design.md#_field_stability)).

> 4. ECS team reviews holistically; on approval they assign the **RFC number** and merge.

Reviewer: this is no true anymore, person should assign next available number

Reviewer: Have you used LLM to generate those? It means we might have somewhere leftovers for this behaviour? We have changed it recently

Author: partly LLM.. It generated the overall content and I tweeked the order and what it should include and what not.

Author: Will update the text..

> 5. Schema landings at agreed maturity may follow in the same or follow-up PRs (per team practice).

Reviewer: I”m not sure what this exactly means? I mean I thought we require people to have both schema andproposal in one PR

Reviewer 2: We want the schema in the RFC folder as supporting info, but the actual additions to the schema may come in the same PR (like Susan's genai one) or after in a subsequent and separate PR (presumable like Ur's entity one)

Author: I think we want the schema and proposal to land in the same PR , if applicable.

Reviewer 2: Ideally yes, but sometimes there is a trickiness to getting the schema implemented 'correctly' due to the subset file https://github.com/elastic/ecs/blob/main/schemas/subsets/main.yml

Author: Yeah.. So restructured the statement to `If applicalbe,....`

| **Summary** | 2–5 sentences: what, why, impact. |
| **Usage** | End-to-end: producer → storage → queries/dashboards/detections. |
| **Fields** | Every proposed field: name, type, description, level/maturity, example. Prefer YAML blocks. If `object`/`flattened` without children, justify shape and conflict avoidance (see [schemas/README.md](../../../schemas/README.md)). |
| **Source data** | ≥2 real examples (JSON/logs); link or place large payloads under `rfcs/text/<n>/`. |
| **Scope of impact** | Ingestion (Beats/Agents), Kibana/apps, ECS repo (docs/tooling). |
| **Concerns** | Risks + **resolved** mitigations; OTel overlap; naming; adoption. |
| **People** | Author, SMEs, reviewers. |
| **References** | Prior art, semconv links, related issues/PRs. |

## `rfcs/text/<number>/` folder

When the RFC adds or changes fields:

- Create **`rfcs/text/<number>/`** with standalone **YAML** snippets, large JSON examples, or mappings — especially when the markdown would be huge.
- Use the **next free** folder number (scan `rfcs/text/`; duplicates get fixed at merge per template notes).
- Align filenames with affected field sets (e.g. `faas.yml`, `gen_ai.yaml`) for reviewer navigation.

> ## OTel alignment (donation period)

Reviewer: this should be changed in both contributing guide and here as well
we don’t do active donation as of now, we do map ecs to otel through tags `match` etc, but I’m not sure it’s a must

- Call out **match**, **equivalent**, **related**, **conflict** explicitly; plan parallel semconv PR if needed (does not need to merge before ECS RFC PR).
- When implementing later in `schemas/*.yml`, each field needs valid `otel:` metadata per [CONTRIBUTING.md](../../../CONTRIBUTING.md).

## Maturity choice (alpha vs beta)

- **Alpha** — earlier, may change more; good for exploratory or fast-moving domains.
- **Beta** — clearer adoption path; still subject to change before GA.
- **Mixture** — some fields alpha, some beta; explain per field or group.

Promotion after merge is **out of band** from the RFC (team process per PROCESS.md).

## PR hygiene

- Link the **Proposal PR** at the bottom of the RFC (`### RFC Pull Requests`).
- Do **not** replace process with old multi-stage labels found in historical RFCs under `rfcs/text/*.md`; those are legacy examples only.

## Example RFCs (depth reference)

- **FaaS + service reuse:** [rfcs/text/0027-faas-fields.md](../../../rfcs/text/0027-faas-fields.md) + folder `rfcs/text/0027/`.
- **GenAI security fields:** [rfcs/text/0050-gen_ai-security-fields.md](../../../rfcs/text/0050-gen_ai-security-fields.md).

> - Use these for **structure** (fields, usage, source data, concerns), not for **stage** metadata.

Reviewer: what stage metadata here means?

Author: The existing RFCs have `Stage` metadata like stage: **Proposal** etc., So mentioning not to copy the stage details directly. Since that is intended to change after the review of RFC , as I understand?

Reviewer: I still don't understand the sentence.
`Use these for... not for stage metadata` - what these? How they should be used?

Reviewer 2: Since we no longer are doing the stages, we don't need the AI to care about that info. But those RFCs are still useful in terms of the info they state

Author: Ok.. I ll take off the stage part..

Author: Fixed in https://github.com/elastic/ecs/pull/2608/commits/067f0e6edffee739a4d68b2fc054b7ddb90799c1

  and optional OTel mapping. Use when a change needs an RFC, when drafting or
  reviewing RFC PRs, or when the user asks how to propose new ECS field sets or
  substantial schema changes.
---

# ECS RFC (Proposal) guide

## When this applies

Use after **ecs-pr-triage** (or equivalent judgment) says **Needs RFC**, or when the user is starting a **new field set**, **breaking** change, **novel use case**, or **ECS-wide** design.

Authoritative process: [rfcs/PROCESS.md](../../../rfcs/PROCESS.md). Template: [rfcs/0000-rfc-template.md](../../../rfcs/0000-rfc-template.md). High-level triggers: [rfcs/README.md](../../../rfcs/README.md).

## Current process (short)

1. Single **Proposal** stage — template must keep `Stage: **Proposal**`.
2. Contributor opens a **PR** that adds the RFC markdown under `rfcs/` (name like `0000-<dash-separated-name>.md` until numbered).
3. Specify **Target maturity:** `alpha`, `beta`, or `mixture` (see [Field stability](../../../docs/reference/ecs-principles-design.md#_field_stability)).
4. The PR author assigns the **next available RFC number** (scan `rfcs/text/` for the highest existing number). ECS team reviews holistically and merges on approval.

> 5. The RFC PR **must include** the schema changes (`schemas/*.yml`, generated artifacts, docs) at the agreed maturity level — proposal and implementation land together in a single PR.

Reviewer: ```suggestion
5. If applicable, the RFC PR **should include** the schema changes (`schemas/*.yml`, generated artifacts, docs) at the agreed maturity level — proposal and implementation land together in a single PR.
```

| **Usage** | End-to-end: producer → storage → queries/dashboards/detections. |
| **Fields** | Every proposed field: name, type, description, level/maturity, example. Prefer YAML blocks. If `object`/`flattened` without children, justify shape and conflict avoidance (see [schemas/README.md](../../../schemas/README.md)). |
| **Source data** | ≥2 real examples (JSON/logs); link or place large payloads under `rfcs/text/<n>/`. |
| **Scope of impact** | Ingestion (Beats/Agents), Kibana/apps, ECS repo (docs/tooling). |
| **Concerns** | Risks + **resolved** mitigations; OTel overlap; naming; adoption. |
| **People** | Author, SMEs, reviewers. |
| **References** | Prior art, semconv links, related issues/PRs. |

## `rfcs/text/<number>/` folder

When the RFC adds or changes fields:

- Create **`rfcs/text/<number>/`** with standalone **YAML** snippets, large JSON examples, or mappings — especially when the markdown would be huge.
- Use the **next free** folder number (scan `rfcs/text/`; duplicates get fixed at merge per template notes).
- Align filenames with affected field sets (e.g. `faas.yml`, `gen_ai.yaml`) for reviewer navigation.

## OTel alignment (optional)

> ECS does not actively donate fields to OpenTelemetry Semantic Conventions at this time. However, ECS tracks relationships between its fields and OTel semconv via `otel:` metadata tags in `schemas/*.yml`.

Reviewer: ```suggestion
ECS tracks relationships between its fields and OTel semconv via `otel:` metadata tags in `schemas/*.yml`.
```

- The `otel:` metadata is used to generate alignment documentation — it is not a gate for merging.
- If unsure whether an OTel mapping applies, omit it; it can be added later.

## Maturity choice (alpha vs beta)

- **Alpha** — earlier, may change more; good for exploratory or fast-moving domains.
- **Beta** — clearer adoption path; still subject to change before GA.
- **Mixture** — some fields alpha, some beta; explain per field or group.

Promotion after merge is **out of band** from the RFC (team process per PROCESS.md).

## PR hygiene

- Link the **Proposal PR** at the bottom of the RFC (`### RFC Pull Requests`).
- Do **not** replace process with old multi-stage labels found in historical RFCs under `rfcs/text/*.md`; those are legacy examples only.

## Example RFCs (depth reference)

- **FaaS + service reuse:** [rfcs/text/0027-faas-fields.md](../../../rfcs/text/0027-faas-fields.md) + folder `rfcs/text/0027/`.
- **GenAI security fields:** [rfcs/text/0050-gen_ai-security-fields.md](../../../rfcs/text/0050-gen_ai-security-fields.md).
- Use these for **structure** (fields, usage, source data, concerns), not for **stage** metadata.

## Implementation in the RFC PR

> The RFC PR itself must include the schema implementation: `schemas/*.yml` changes, `make`-generated artifacts, and a **CHANGELOG.next.md** entry. There is no separate "handoff" — proposal and schema land together.

Reviewer: ```suggestion
The RFC PR itself should include the schema implementation: `schemas/*.yml` changes, `make`-generated artifacts, and a **CHANGELOG.next.md** entry. There is no separate "handoff" — proposal and schema land together.
```
